File: python/0322零钱兑换.py

# 组合完全背包
def coinChange(coins: list, amount: int) -> int:
    
    dp=[10e4+1]*(amount+1)
    dp[0]=0

    for c in coins:
        for j in range(c,amount+1):
            dp[j]=min(dp[j],dp[j-c]+1)
        
    return dp[-1] if dp[-1]!=10e4+1 else -1

print(coinChange([2],3))

File: csharp/0322零钱兑换.cs
public partial class Solution
{
    public int CoinChange(int[] coins, int amount)
    {
        int[] dp = new int[amount + 1];

        for (int i = 1; i<=amount;i++)
        {
            dp[i] = (int)(10e4 + 1);
        }

        foreach(int c in coins)
        {
            for (int j = c; j<=amount;j++)
            {
                dp[j] = Math.Min(dp[j], dp[j - c] + 1);
            }
        }

        return dp[amount] == (int)(10e4 + 1) ? -1 : dp[amount];
    }
}